Admin tools: Adds a forceMove tool to build mode

Miraviel opened the pull request:
## What Does This PR Do

Adds a new option to build mode called "forcemove". Right-click on the atom you want to move, left click the destination.

<sup>(It also fixes a stray pixel on the buildmode atmos icon)</sup>

## Why It's Good For The Game

We frequently move things with `forceMove()`, it is something every new admin is taught how to do. However, it is incredibly tedious.

One usage example for this is filling crates with big objects - when we try to send an engine crate or something syndie researchers requested via fax, we need to `right-click -> VV -> mark datum` the crate, then `right-click -> proccall -> forceMove() -> one parameter -> marked datum` every single object we want to put in. It is a soul-killing process.

This tool also prevents you to do silly things like moving mobs into effects **(EDIT: or ghosts)** by accident. If we have any other "please for the love of god don't move this or move into this" type of restrictions, let me know, I could not come up with anything else.

## Images of changes

1. Right-clicking the ? mark shows an explanation
2. Error message when you try to move things you shouldn't
3. Example of moving big stuff into crates
4. Example of moving mobs into crates

https://user-images.githubusercontent.com/33333517/210213524-1fbab8c7-fcaa-468d-837d-688312735f51.mp4

Edit: it has a green selection overlay now as well:

https://user-images.githubusercontent.com/33333517/212074998-037b2198-acce-497f-b6df-3dbf68be17aa.mp4

## Testing

Did the video, properly crashed my server first because I forgot a nullcheck.

## Changelog
:cl:
add: Admin tools - Added a forceMove tool to build mode.
/:cl:

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,52 @@
+		log_admin("Build Mode: [key_name(user)] forcemoved [selected_atom] to [destination] at ([destination.x],[destination.y],[destination.z]).")
+
+		selected_atom = null
+		return
```
```suggestion
```
dangling return

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,52 @@
+		to_chat(user, "<span class='notice'>'[selected_atom]' is selected to be moved.</span>")
+
+	// Selecting the destination to move to
+	if(left_click)
```
make this an early return please

Miraviel replied:
Resolved the other two! Not sure what you refer to here - I find it identical to your third point?

Sirryan2002 replied:
```dm
if(left_click)
	return
```
then unindent all the code below

AffectedArc07 replied:
Im gonna put a bit of input in here. Honestly I think this is better from a readability POV.

Consider the following
```dm
if(left)
    l_stuff()

if(right)
    r_stuff()

if(mid)
    m_stuff()
```

whereas 

```dm
if(left)
    l_stuff()

if(right)
    r_stuff()

if(!mid)
    return

m_stuff()
```

Just looks off.

---

In an ideal world we would have 
```dm
switch(click_type)
    if(LEFT_CLICK)
        l_stuff()

    if(RIGHT_CLICK)
        r_stuff()

    if(MID_CLICK)
        m_stuff()
```
Possible future refactor?

Miraviel replied:
Resolved in c5ce5a6

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,52 @@
+			return
+
+		selected_atom = A
+		to_chat(user, "<span class='notice'>'[selected_atom]' is selected to be moved.</span>")
```
```suggestion
		to_chat(user, "<span class='notice'>'[selected_atom]' is selected to be moved.</span>")
		return
```
the code below this will not be called in the current logic if `right_click` is true, just return here

Miraviel commented:
@Sirryan2002 The code now also checks for observers, I just noticed it is too easy to accidentally forcemove things into zooming observers - please take a look at it.

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,55 @@
+		return
+
+	// Block these as they can only lead to issues
+	if(iseffect(destination) || isobserver(destination))
```
instead of doing `istype()` checks here. Can you use `is_type_in_list()` and store a `var/static/list` of types that are banned for these datums.

This way its really easy to add more blacklisted things down the line (when an admin inevitably breaks something with this). Also here's a few more types that need to be added to the blacklist

- All types of `/obj/machinery/atmospherics` that are currently anchored <--- forcemoving these while anchored may break pipenets
- all types of `/obj/structure/disposalpipe` that are anchored, same reason for above
- all types of `/obj/machinery/power` that are anchored <---- should not be moved off of wire nodes via forcemove
- `/obj/structure/cable` <--- Shouldn't be moved off a turf or into anything via forcemove
- `/obj/machinery/field/containment` <--- should not be moved via forcemove

Either they need to go in the blacklist or they need special behavior overrides for `forcemove()` to disconnect them from station systems properly.

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,55 @@
+		to_chat(user, "<span class='danger'>You should not move atoms into effects or ghosts.</span>")
+		return
+
+	// Typecasting for the right forceMove()
+	if(ismob(selected_atom))
+		var/mob/M = selected_atom
+		M.forceMove(destination)
+	else
+		var/atom/movable/AM = selected_atom
+		AM.forceMove(destination)
```
it should use the right `forceMove()` without typecast ? `mob` types override the `forceMove()` proc already

Sirryan2002 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,55 @@
+// Once the atom is moved, its selection is cleared to avoid any further moves by accident.
+/datum/buildmode_mode/forcemove
+	key = "forcemove"
+	var/atom/selected_atom		// The atom we want to move
```
this needs to have a signal registered on it to ensure that references are properly cleared if the atom is `qdel`'d

AffectedArc07 replied:
Or just a soft ref

farie82 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,55 @@
+			to_chat(user, "<span class='danger'>You cannot move effects, turfs or areas.</span>")
+			return
+
+		selected_atom = A
```
Echoing Sirryans worries. But I think adding an overlay to the item, only visible to the user ofc, would be a better solution.

Most other admin tools already break the game if used incorrectly. But an overlay with this feature would help alleviate the misclicks at least.

Miraviel commented:
@Sirryan2002 @farie82 
1. Removed the unnecessary typecasting
2. Added signals
3. Added a green overlay 


https://user-images.githubusercontent.com/33333517/212074272-7449b28f-6cd2-498e-8dcf-d90c7ae7339e.mp4

SteelSlayer commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,77 @@
+	selected_overlay = null
+
+// If it gets deleted mid-movement, remove the overlay and its attachment to the forcemove tool
+/datum/buildmode_mode/forcemove/proc/on_selected_atom_deleted()
+	UnregisterSignal(selected_atom, COMSIG_PARENT_QDELETING)
+	selected_atom = null
```
For consistency
```suggestion
/datum/buildmode_mode/forcemove/proc/on_selected_atom_deleted()
	SIGNAL_HANDLER
	UnregisterSignal(selected_atom, COMSIG_PARENT_QDELETING)
	selected_atom = null
```

SteelSlayer reviewed (commented):
One thing, looks good otherwise.

farie82 commented on code/modules/buildmode/submodes/forcemove.dm:
```diff
@@ -0,0 +1,78 @@
+			return
+
+		// If we had something previously selected, handle its signal and overlay
+		if(selected_atom)
+			UnregisterSignal(selected_atom, COMSIG_PARENT_QDELETING)
+
+		if(selected_overlay)
+			remove_selected_overlay(user)
```
This should happen the same time all the time right? So even when the item gets deleted it should also "unselect"  the overlay so it goes away. Basically the overlay shouldn't exist if the atom doesn't exist anymore

Miraviel replied:
There is for a very edge case where the atom gets destroyed after selecting it. In that case, the signal attached to the atom will make `selected_atom` null, but the `selected_overlay` stays as is (as I cannot handle `user.client.images` via the destroy signal).
